.github/ISSUE_TEMPLATE/feature_request.yml

```diff
@@ -54,10 +54,13 @@ body:
 
                   while (window.isOpen())
                   {
-                      while (const auto event = window.pollEvent())
+                      while (const std::optional event = window.pollEvent())
                       {
-                          if (event.is<sf::Event::Closed>())
+                          if (event->is<sf::Event::Closed>())
+                          {
                               window.close();
+                              break;
```

Comment:
I suppose this `break` ensures that event processing ceases when a close is requested? In theory this reduces the latency between clicking the close button the application actually closing. Is that your intent?

Reply:
That's not my intent, but it's a nice side benefit of the change.

The intention is to prevent event processing from causing issues when using `waitEvent`, as `waitEvent` now asserts `m_impl != nullptr`, but `WindowBase::close` immediately invokes `m_impl.reset()`.

We need to call `break;` to avoid accessing a deleted `m_impl` on the next iteration.

Now that I wrote this, it does sound like a subtle trap if the `break;` is forgotten... But the only way I can think of fixing this is to revert `waitEvent` to return a `std::optional<sf::Event>`, which loses a nice API improvement that this PR would provide.

I think a decent alternative would be to do this in `waitEvent`:

```cpp
Event WindowBase::waitEvent()
{
    if (m_impl == nullptr)
    {
        sf::err() << "waitEvent invoked on destroyed window -- did you forget a `break`?" << std::endl;
        std::terminate();
    }

    std::optional<sf::Event> event = m_impl->popEvent(true /* blocking */);
    assert(event.has_value());

    filterEvent(*event);
    return *event;
}
```

Reply:
I've added the additional safety measure to `waitEvent` and updated its documentation.

Reply:
> We've found an edge case where `waitEvent` can return an empty state

It's more of a precondition failure -- the precondition is that `waitEvent` should not be called  on a window that's already been closed. Returning a `sf::Event::Closed` is a way of not terminating nor getting UB, but it's not behavior the user should rely on.

Perhaps I can add an `err()` before the return to make it clear that the user is doing something wrong and we're just being nice?

Reply:
Actually, this just reminded me of the longstanding issue of `waitEvent` never returning if there are no more events to be produced. For example if you run `waitEvent` in a separate thread and in the other thread close the window. The event thread will continuing spinning as it doesn't get any events to process: https://github.com/orgs/SFML/projects/4?pane=issue&itemId=23482732# / https://github.com/SFML/SFML/issues/1184

Since SFML is anyways doing a spinning while loop, due to Joystick event processing, I'd actually propose to change `waitEvent` to take a custom timeout parameter. Blocking methods should always have a way to cancellation token or timeout option.

In that case you'll have another situation where waitEvent has a "failure case".

The user isn't doing anything wrong, when calling any of the window methods after closing the window. Because you still have a valid object and as such any operation on that object should remain valid for the lifetime of the object.

Reply:
> Actually, this just reminded me of the longstanding issue of `waitEvent` never returning if there no more events are being produced. For example if you run `waitEvent` in a separate thread and in the other thread closes the window. The event thread will continuing spinning as it doesn't get any events to process: https://github.com/orgs/SFML/projects/4?pane=issue&itemId=23482732# / #1184
> 
> Since SFML is anyways doing a spinning while loop, due to Joystick event processing, I'd actually propose to change `waitEvent` to take a custom timeout parameter. Blocking methods should always have a way a cancellation token or timeout option.
> 
> In that case you'll have another situation where waitEvent has a "failure case".
> 
> The user isn't doing anything wrong, when calling any of the window methods after closing the window. Because you still have a valid object and as such any operation on that object should remain valid for the lifetime of the object.

I like the timeout idea, I will work on a PR for that specifically (against master) and then integrate that into this one as they are separate concerns.

Reply:
Created PR for the timeout here: https://github.com/SFML/SFML/pull/3094

Reply:
> It's more of a precondition failure -- the precondition is that waitEvent should not be called on a window that's already been closed.

In my opinion calling `waitEvent` on a closed window is precondition violation. We should log an error, `assert(false)`, then return `sf::Event::Closed` to ensure behavior is defined in release builds. I still want that function to return an `sf::Event` rather than `std::optional<sf::Event>`.

Reply:
I strongly disagree. As long as an object was initialized, it's in a valid state. We cannot suddenly have functions semi-fail because we assign random preconditions.

Besides, with the timeout proposal, you actually get an expected failure case.

I'm also against adding `break;` everywhere.

Reply:
My views:

- It's very unfortunate that we lose the nice non-`optional` API for `waitEvent` if we add the timeout, but I think it's a good idea to have one;

- I agree with @ChrisThrasher that calling `waitEvent` on a closed window should be a precondition violation and I agree with his proposed solution -- however if we do add a timeout it would make sense to return `std::nullopt` there as well;

- I disagree with both of you regarding `break;` -- I think that the only sensible thing to do when you close a window is to break out of the event handling loop, and I'd like to promote that practice regardless of whether or not it's strictly necessary for the correctness of the program.

Reply:
I think adding `break` in all these places is unrelated to the core mission of this PR. If you want to make the case for it, I'd rather that happen in a separate PR because discussing it is stalling the more important API changes. In general it gets back to a personal preference of mine that PRs not try to solve multiple orthogonal problems at once.

As for `waitEvent` changes, does that PR need to be settled first before this PR's review can continue? Help us understand the best order of operations for all the PRs you have recently opened.

Reply:
#3094 was merged in, I've rebased this PR on that. I've also merged in #3096 into this PR, I would like to merge that as a prerequisite but I can extract it out if required. Resolving for now.